### ISSUE TITLE
 345.464 Error ModManager.cpp:1625: Reached id limit for entity. The game can't contain more than 65535 instances of this prototype due to hardcoded limits of the engine. See the log file for more detail.

### ISSUE THREAD

**Issue by Leo0806-studios** [millerscout/MMD-Factorio-Mods#3]
[factorio-current.log](https://github.com/millerscout/MMD-Factorio-Mods/files/12840636/factorio-current.log)

**Comment by Leo0806-studios** [comment]
it apperantly creates a new prototype for every building for every level.
with heavily modded setups a defult value of 100 is to high

**Comment by millerscout** [comment]
Agreed with you heavily modded games will teach that problem, currently I'm not fixing that because I need to plan alternatives.


Later on I'll try to share with you the alternatives

**Comment by malomaster** [comment]
Maybe you can try [my idea](https://github.com/renoth/factorio-factory-levels/issues/40) I posted for Factory levels?

**Comment by millerscout** [comment]
Yeah I read that, it's a good solution, I was trying to figure out a way, but it takes time to find a proper solution for it. Lots of research, and my daughter draws my attention from it,  on my free time I'll try make something about it

**Comment by millerscout** [comment]
Hey, I just had a way to make it work, it'll take lake 2/3 days to write, but when it completes may you help me checking?

**Comment by Leo0806-studios** [comment]
> Hey, I just had a way to make it work, it'll take lake 2/3 days to write, but when it completes may you help me checking?

if its just play and stress testing, sure but for more i don't know enough about lua

**Comment by millerscout** [comment]
Lua Im quite confident, stress test + balance is something hard to do 🫣

**Comment by millerscout** [comment]
![Screenshot 2023-10-08 233828](https://github.com/millerscout/MMD-Factorio-Mods/assets/1369224/a81c4da4-4d75-43b7-8910-5db92ed9aed8)

@malomaster @Leo0806-studios  i'm not a proper designer, BUT!
i think it might work

**Comment by malomaster** [comment]
If this is just an invisible (hidden) modules, then it's fine. But, if those icons will be seen as machine levels, then all of them need to redesigned. Maybe it will be better if icons will be something like rank icons in games like Call of Duty, World of Tanks and many other similar games with ranks.
But it can be postponed, I think icon design is not the most important now, the first thing is how it will work. Are level modules' bonuses hardcoded and applied to all regardless of machine type and base stats, or there will be 100 different modules for each machine type, or, like I suggested, mod generates a table with bonuses for each assembling entity type (stone furnace, steel furnace, assembler 1, assembler 2 etc.), and there are only 100 level modules, but bonuses they give vary depending on entity type?

**Comment by millerscout** [comment]
it's not possible yet to make "invisible modules", there's a way to do it but it takes alot of effort, on the other hand the current approach is very "pleasing" to hover a building or Alt-view  buildings and visually get a feedback about the level.


i designed the module using multiple sprites, its possible to write the level until 999 dynamically

as it stands right now the modules are hardcoded, bonuses are shared over multiple types.

i don't know if there's a real benefit making more modules yet.

on alpha version if you got like 50 entities turns out it generates 5K prototypes, for users that uses pymods and other big mods, the load time increases dramatically + it may trigger the 65k prototype limit.

on 1.0, using the module approach it uses just 100.

but for a 1.1 version we may improve detecting tiers between entities and cap the level accordingly

and maybe implement this idea: https://mods.factorio.com/mod/xp-for-buildings/discussion/65235d694dfe6f28cbfc19e5

and something like below, could be possible:
if there's two assemblies
1st assembly - 0.5 crafting-speed
2nd assembly - 3 crafting-speed


the 1st assembly could level up until the crafting speed reaches the 2nd assembly
if the crafting-multiplier is 0.2 the max level for 1st assembly is 13/14

**Comment by millerscout** [comment]
now we have another problem

as i'm testing 

using modules we have flat increase in % it's broadly limited 

we cannot increase an entity speed by .001% only works using integers, 

maybe it's not the best approach 
https://lua-api.factorio.com/latest/concepts.html#ModuleEffectValue

**Comment by millerscout** [comment]
[zzz-xp-for-buildings.zip](https://github.com/millerscout/MMD-Factorio-Mods/files/12851872/zzz-xp-for-buildings.zip)
if you want to check how's working

**Comment by DrWiFi** [comment]
Would it be possible to move the modules into a separate tab group? Like with Schall's Circuits?

**Comment by malomaster** [comment]
> using modules we have flat increase in % it's broadly limited
> we cannot increase an entity speed by .001% only works using integers,

lua-api says: `pollution={bonus=0.075}`
Isn't it means that module has 7.5% pollution increase? I think a fraction of 0.1% will be enough for our purposes. I don't think it makes sense to split into smaller values — in the end, you can just use math.floor/ceil to round up to 0.1%. Or to 1%, if the game understands only integer percents. At least it will be more readable.

**Comment by millerscout** [comment]
> Would it be possible to move the modules into a separate tab group? Like with Schall's Circuits?

I think so, need to check how he made that

**Comment by millerscout** [comment]
> > using modules we have flat increase in % it's broadly limited
> > we cannot increase an entity speed by .001% only works using integers,
> 
> lua-api says: `pollution={bonus=0.075}` Isn't it means that module has 7.5% pollution increase? I think a fraction of 0.1% will be enough for our purposes. I don't think it makes sense to split into smaller values — in the end, you can just use math.floor/ceil to round up to 0.1%. Or to 1%, if the game understands only integer percents. At least it will be more readable.

Yes, but in that case alien modules does the same thing, I like factory levels because of its granularity.

For now I'll stick to prototypes approach,
There's a lot of room for optimization.

**Comment by malomaster** [comment]
> I think so, need to check how he made that

You also can download 5dim_core and, for example, 5dim_battlefield mod, in "core" there are very easy-to-read samples how to make tab groups. Just copy and rename any group and subgroup and then add this subgroup to module recipes subgroup.

> Yes, but in that case alien modules does the same thing, I like factory levels because of its granularity.

Me too, but, if module levels method will be a lot faster and/or use less cpu resources than entity levels method, but will have less accuracy (0.1% crafting speed/productivity/pollution/energy usage/attack speed/damage bonus etc.), with math.floor/ceil you can just use something like `productivity_bonus = math.floor(machine.level / 3 * 2) * 0.1 %`. With formula like this, machine will get bonus every second and third level out of three. Or it may be 1% every 2, 3, 5 or even 10 levels – why not? It will be like in some rpg games – like, on 2nd lvl you got 1 Str and 1 Dex, on 3rd -- 1 Dex and 1 Con, on 4th -- 1 Str and 1 Int etc. Although I doubt that we will really need values less than 0.1%. Also, it hardly makes sense to give players possibility to set max_level greater than 100, except the last tier machines -- they can have maximum of, say, 200, 300 or even 500 levels if there are not so much machine entity types.
